pkg/registry/certificates/certificates/storage/metrics.go

```diff
@@ -0,0 +1,170 @@
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"signerName"},
+	)
+)
+
+func init() {
+	registerMetricsOnce.Do(func() {
+		legacyregistry.MustRegister(csrDurationRequested)
+		legacyregistry.MustRegister(csrDurationHonored)
+	})
+}
+
+var registerMetricsOnce sync.Once
+
+func TestCollectMetrics(ch chan<- prometheus.Metric) {
+	csrDurationRequested.Collect(ch)
+	csrDurationHonored.Collect(ch)
+}
+
+type counterVecMetric interface {
+	WithLabelValues(...string) metrics.CounterMetric
+}
+
+func countCSRDurationMetric(requested, honored counterVecMetric) genericregistry.BeginUpdateFunc {
+	return func(ctx context.Context, obj runtime.Object, old runtime.Object, options *metav1.UpdateOptions) (genericregistry.FinishFunc, error) {
+		return func(ctx context.Context, success bool) {
+			if !success {
+				return // ignore failures
+			}
+
+			if dryrun.IsDryRun(options.DryRun) {
+				return // ignore things that would not get persisted
+			}
+
+			if !utilfeature.DefaultFeatureGate.Enabled(features.CSRDuration) {
+				return
+			}
+
+			oldCSR := old.(*certificates.CertificateSigningRequest)
+
+			// if the old CSR already has a certificate, do not double count it
+			if len(oldCSR.Status.Certificate) > 0 {
+				return
+			}
+
+			if oldCSR.Spec.ExpirationSeconds == nil {
+				return // ignore CSRs that are not using the CSR duration feature
+			}
+
+			issuedCert := obj.(*certificates.CertificateSigningRequest).Status.Certificate
+
+			// new CSR has no issued certificate yet so do not count it.
+			// note that this means that we will ignore CSRs that set a duration
+			// but never get approved/signed.  this is fine because the point
+			// of these metrics is to understand if the duration is honored
+			// by the signer.  we are not checking the behavior of the approver.
+			if len(issuedCert) == 0 {
+				return
+			}
+
+			signer := compressSignerName(oldCSR.Spec.SignerName)
+
+			// at this point we know that this CSR is going to be persisted and
+			// the cert was just issued and the client requested a duration
+			requested.WithLabelValues(signer).Inc()
+
+			certs, err := cert.ParseCertsPEM(issuedCert)
+			if err != nil {
+				utilruntime.HandleError(fmt.Errorf("metrics recording failed to parse certificate for CSR %s: %w", oldCSR.Name, err))
+				return
+			}
+
+			// now we check to see if the signer honored the requested duration
+			certificate := certs[0]
+			wantDuration := csr.ExpirationSecondsToDuration(*oldCSR.Spec.ExpirationSeconds)
+			actualDuration := certificate.NotAfter.Sub(certificate.NotBefore)
+			if isDurationHonored(wantDuration, actualDuration) {
+				honored.WithLabelValues(signer).Inc()
+			}
+		}, nil
+	}
+}
+
+func isDurationHonored(want time.Duration, got time.Duration) bool {
+	delta := want - got
+	if delta < 0 {
+		delta = -delta
+	}
+
+	// short-lived cert backdating + 5% of want
+	// TODO should we have an upper limit on the 5%?
+	maxDelta := 5*time.Minute + (want / 20)
```

Comment:
I see the 5 minutes from our hard-coded backdate... where does the 5% come from?

Reply:
I was trying to come up with some fudge factor to allow for external signers that may do like a 10 minute backdate or something.  I can drop it all together if you want...